tensorflow-framework/src/main/java/org/tensorflow/framework/losses/Losses.java

```diff
@@ -0,0 +1,711 @@
+package org.tensorflow.framework.losses;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.impl.LossTuple;
+import org.tensorflow.framework.losses.impl.LossesImpl;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.ReduceAll;
+import org.tensorflow.op.core.ReduceMax;
+import org.tensorflow.op.core.ReduceSum;
+import org.tensorflow.op.math.Mean;
+import org.tensorflow.op.math.Softplus;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.TInt64;
+import org.tensorflow.types.family.TNumber;
+
+/** Built-in loss functions. */
+public class Losses {
```

Comment:
What is the reason for having a seperate `Losses` class that implements all individual loss algorithms instead of implementing them directly in their respective class?

I guess is to offer the choice to the user to choose an object-oriented API or a functional one? Then if we offer this, should we do the same for all other concepts in the framework?

I wondering if we shouldn't take that decision for our users and only present a single API to accomplish a given task, for simplicity and consistency. I personally have a preference for the OO approach.

Reply:
Metrics class will have to call these methods directly. The handling of the `reduction` is different in metrics from losses.

Reply:
Also, the methods in `Losses`, do not apply a `reduction`. `Loss` and `Metric` handle the reduction differently, so the raw `loss` needs to be fetched first. If we didn't have the `loss` methods , the the metric classes would have to duplicate that functionality. Also, I can only assume that Keras has this method split from the class for some reason. Of course, we could move the static method to its corresponding class and leave it static so that it can be still invoked independent of any reduction logic.

It is worth noting, that not all `Metrics` are based on `Loss` algorithms, such as `AUC` (Area under the curve).

Also, there is one class,
`CosineSimilarity`, that uses a different algorithm between the Loss class (cosineSimilarity) as opposed to the corresponding Metric class (cosineProximity). While these two methods are in the same category, they do not compute the exact same result. I tried to reasearch why there is a difference, maybe @Craigacp can shed some light on this. One of the questions in `Metrics` is should the metric class be renamed to CosineProximity.

Reply:
That's because the losses version of cosine similarity isn't actually cosine similarity, it's the negative of it. Because the cosine of 0 is 1, but here when the two vectors are pointing in the same direction it emits -1. The metrics one actually computes cosine similarity.

I assume it's to make the minimize function work consistently.